radicle-cob/src/tests.rs

```diff
@@ -0,0 +1,253 @@
+    let neil_signer = gen::<MockSigner>(2);
+    let neil = test::Person::new(&storage, "gaiman", *neil_signer.public_key()).unwrap();
+    let terry_signer = gen::<MockSigner>(1);
+    let terry = test::Person::new(&storage, "pratchett", *terry_signer.public_key()).unwrap();
+    let proj = test::Project::new(&storage, "discworld", *terry_signer.public_key()).unwrap();
+    let terry_proj = test::RemoteProject {
+        project: proj.clone(),
+        person: terry.clone(),
+    };
+    let neil_proj = test::RemoteProject {
+        project: proj,
+        person: neil.clone(),
+    };
+    let typename = "xyz.rad.issue".parse::<TypeName>().unwrap();
+    let cob = create(
+        &storage,
+        terry_signer,
+        &terry_proj,
+        Create {
+            author: Some(terry),
+            contents: history::Contents::Automerge(b"issue 1".to_vec()),
+            typename: typename.clone(),
+            message: None,
+        },
+    )
+    .unwrap();
+    copy_to(
+        storage.as_raw(),
+        &terry_proj,
+        &neil_proj,
+        &typename,
+        *cob.id(),
+    )
+    .unwrap();
+
+    let updated = update(
+        &storage,
+        neil_signer,
+        &neil_proj,
+        Update {
+            author: Some(neil),
+            changes: history::Contents::Automerge(b"issue 2".to_vec()),
+            object_id: *cob.id(),
+            typename,
+            message: None,
+        },
+    )
+    .unwrap();
+
+    // traverse over the history and filter by changes that were only authorized by terry
+    let contents = updated.history().traverse(Vec::new(), |mut acc, entry| {
+        let author = match entry.author() {
+            Some(author) => test::Person::find_by_oid(storage.as_raw(), *author).unwrap(),
+            None => None,
+        };
+        let project = test::Project::find_by_oid(storage.as_raw(), entry.resource()).unwrap();
+
+        if let (Some(author), Some(project)) = (author, project) {
+            if project.delegate_check(&author) {
+                acc.push(entry.contents().as_ref().to_vec())
+            }
+        }
+        ControlFlow::Continue(acc)
+    });
+
+    assert_eq!(contents, vec![b"issue 1".to_vec()]);
+
+    // traverse over the history and filter by changes that were only authorized by neil
+    let contents = updated.history().traverse(Vec::new(), |mut acc, entry| {
+        acc.push(entry.contents().as_ref().to_vec());
+        ControlFlow::Continue(acc)
+    });
+
+    assert_eq!(contents, vec![b"issue 1".to_vec(), b"issue 2".to_vec()]);
+}
+
+fn gen<T: Arbitrary>(size: usize) -> T {
+    let mut gen = quickcheck::Gen::new(size);
+
+    T::arbitrary(&mut gen)
+}
+
+fn copy_to(
+    repo: &git2::Repository,
+    from: &test::RemoteProject,
+    to: &test::RemoteProject,
+    typename: &TypeName,
+    object: ObjectId,
+) -> Result<(), git2::Error> {
+    let original = {
+        let name = format!(
+            "refs/rad/{}/cobs/{}/{}",
```

Comment:
I'm probably missing something, but what are these refs used for? Looks unfamiliar to me.

Reply:
I had to make up a storage layout for the tests so I just have them under:
```
refs/rad/<project>/<person>/cobs/...
```

These can be whatever you want in `radicle` :kissing_closed_eyes:

Reply:
Oh I see!